### PR TITLE
Add the common module with XDS circuit breaker and rate limiting function

### LINKED ISSUE #1515: Supports xDS protocol to communicate with the Control Plane of Service Mesh
# Value
Sermant is a Proxyless Service Mesh Solution with better performance.The support for the xDS protocol enables Sermant to better integrate into the Cloud Native Ecosystem. In Istio, Sermant can replace Envoy as the data plane to obtain directly the configuration of the control plane, and avoid performance loss and architecture complexity caused by Sidecar containers.
# Objective
Sermant can connect to the control plane of Istio through the xDS API interface to obtain the configurations of CDS, EDS, LDS, and RDS, and parse configurations to data model of the Sermant. Sermant will  implement gradually capabilities such as service discovery , route configuration and flow control, and the the host microservice can implements the Proxyless Mesh capability by mounting Sermant to  invoke directly the upstream service registered with the control plane, such as istiod.
# RoadMap
**v2.0.0 (2024/06/30)**
Supports xDS protocol for service discovery
**v2.1.0 (2024/09/30)**
Supports xDS protocol for routing and load balancing
**v2.2.0 (2024/12/30)**
Supports xDS protocol for flow control

### PR DESCRIPTION
**Pull request by hanbingleixue**
**What type of PR is this?**

Feature

**What this PR does / why we need it?**

Add the common part of Xds flow control plugin

**Which issue(s) this PR fixes？**

Fixes #1515

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * XdsMatchManager, Get BusinessEntity based on xDS routing rules
+ *
+ * @author daizhenyu
+ * @since 2024-08-29
+ **/
+public enum XdsMatchManager {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java]: Specific matching scenarios, such as routing

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            return scenario;
+        }
+        XdsRoute matchedRoute = matchedRouteOptional.get();
+        XdsRouteAction routeAction = matchedRoute.getRouteAction();
+        String cluster = routeAction.getCluster();
+        if (routeAction.isWeighted()) {
+            cluster = selectClusterByWeight(routeAction.getWeightedClusters());
+        }
+        scenario.setRouteName(matchedRoute.getName());
+        scenario.setClusterName(cluster);
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java]: extract as a method, which is routing.

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+ * Scenario information for flow control, primarily used to retrieve matching flow control rules
+ *
+ * @author zhp
+ * @since 2024-11-27
+ */
+public class FlowControlScenario {
+    /**
+     * matched service scenario name
+     */
+    private Set<String> businessNames;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java]: optimize name, what is businessName?

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/handler/AbstractRequestHandler.java [hunk 4]
```diff
@@ -70,10 +71,24 @@ public List<H> getHandlers(RequestEntity request) {
...
+     * gets the specified request handler
+     *
+     * @param flowControlScenario matched scenario information
+     * @return handler
+     */
+    public List<H> getXdsHandlers(FlowControlScenario flowControlScenario) {
+        String businessName = flowControlScenario.getServiceName() + CommonConst.CONNECT
+                + flowControlScenario.getRouteName() + CommonConst.CONNECT + flowControlScenario.getClusterName();
+        Optional<H> handlerOptions = handlers.computeIfAbsent(businessName, fn ->
+                createProcessor(flowControlScenario, businessName));
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/handler/AbstractRequestHandler.java]: createHandler

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    /**
+     * increment Active Request
+     *
+     * @param serviceName service name
+     * @param clusterName route name
+     * @param address request address
+     * @return active request num
+     */
+    public static int incrementActiveRequests(String serviceName, String clusterName, String address) {
+        AtomicInteger activeRequest = getActiveRequestCount(serviceName, clusterName, address);
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: no need temp variable

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Circuit Breaker information
+ *
+ * @author zhp
+ * @since 2024-12-02
+ */
+public class XdsCircuitBreakerInfo {
+    private List<Long> localFailure;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: use deque

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        if (code != 0 && GATE_WAY_FAILURE.contains(String.valueOf(code))) {
+            recordRequestTime(circuitBreakerInfo.getGateWayFailure(), circuitBreakers.getConsecutiveGatewayFailure(),
+                    currentTime);
+        }
+        recordRequestTime(circuitBreakerInfo.getServerFailure(), circuitBreakers.getConsecutive5xxFailure(),
+                currentTime);
+    }
+
+    private static void recordRequestTime(List<Long> times, int count, long currentTime) {
+        if (count == 0) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: count -> failureRequestThreshold

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    public void setActiveRequests(AtomicInteger activeRequests) {
+        this.activeRequests = activeRequests;
+    }
+
+    /**
+     * reset the data
+     */
+    public void cleanRequestDate() {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: Data

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+    private List<Long> serverFailure;
+
+    private boolean isOpen;
+
+    private long circuitBreakerTime;
+
+    private AtomicInteger circuitBreakerCount;
+
+    private AtomicInteger activeRequests;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: remove activeRequests

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+public class XdsCircuitBreakerInfo {
+    private List<Long> localFailure;
+
+    private List<Long> gateWayFailure;
+
+    private List<Long> serverFailure;
+
+    private boolean isOpen;
+
+    private long circuitBreakerTime;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: circuitBreakerEndTime

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * reset the data
+     */
+    public void cleanRequestDate() {
+        this.localFailure.clear();
+        this.gateWayFailure.clear();
+        this.serverFailure.clear();
+        this.activeRequests.set(0);
+        this.circuitBreakerCount.set(0);
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: do not clear circuitBreakerCount

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+    public static void countFailureRequest(FlowControlScenario scenarioInfo, String address, int code,
+            XdsInstanceCircuitBreakers circuitBreakers) {
+        XdsCircuitBreakerInfo circuitBreakerInfo = getCircuitBreakerInfo(scenarioInfo.getServiceName(),
+                scenarioInfo.getRouteName(), address);
+        if (isCircuitBreakerOpen(circuitBreakerInfo, circuitBreakers.getBaseEjectionTime())) {
+            return;
+        }
+        long currentTime = System.currentTimeMillis();
+        if (!XdsThreadLocalUtil.getSendByteFlag()) {
+            circuitBreakerInfo.getLocalFailure().add(currentTime);
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: remove this line

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+     * add failure request
+     *
+     * @param scenarioInfo scenario information
+     * @param address request address
+     * @param code response code
+     * @param circuitBreakers circuit rule
+     */
+    public static void countFailureRequest(FlowControlScenario scenarioInfo, String address, int code,
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: recordFailureRequest

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+
+    /**
+     * handler circuit breaker
+     *
+     * @param scenarioInfo Flow Control Scenario information
+     * @param address service address
+     * @param circuitBreakers circuit rule
+     * @return Should Circuit Breaker Be Triggered
+     */
+    public static boolean needCircuitBreaker(FlowControlScenario scenarioInfo, String address,
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: Trigger circuit breaker when the instance fails to be invoked.

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+    private XdsLoadBalanceService xdsLoadBalanceService;
+
+    /**
+     * constructor
+     */
+    XdsFlowControlHandler() {
+        Logger logger = LoggerFactory.getLogger();
+        try {
+            XdsCoreService xdsCoreService = ServiceManager.getService(XdsCoreService.class);
+            xdsRouteService = xdsCoreService.getXdsRouteService();
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: judge XdsCoreService is null

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * Xds Flow Control handler
+ *
+ * @author zhp
+ * @since 2024-11-28
+ */
+public enum XdsFlowControlHandler {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: This class is used to obtain the configuration and does not execute the flow control logic. It should not be called the flow control handler.

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java [hunk 17]
```diff
@@ -0,0 +1,273 @@
...
+     *
+     * @param serviceName service name
+     * @param clusterName cluster name
+     * @return Service Instance
+     */
+    public Set<ServiceInstance> getAllServerInstance(String serviceName, String clusterName) {
+        if (StringUtils.isEmpty(serviceName) || StringUtils.isEmpty(clusterName)) {
+            return Collections.emptySet();
+        }
+        if (xdsServiceDiscovery == null) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: Judge first in this method

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+    /**
+     * obtain token
+     *
+     * @param serviceName service name
+     * @param routeName route name
+     * @param tokenBucket token rule info
+     * @return Obtain results
+     */
+    public static synchronized boolean consume(String serviceName, String routeName, XdsTokenBucket tokenBucket) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: method name should be clear.

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Circuit Breaker manager
+ *
+ * @author zhp
+ * @since 2024-12-02
+ */
+public class RateLimitManager {
+    private static final Map<String, Map<String, XdsRateLimitInfo>> RATE_LIMIT_INFO_MAP = new ConcurrentHashMap<>();
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: describe what is the key

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+        return false;
+    }
+
+    /**
+     * Fill in tokens
+     *
+     * @param xdsRateLimitInfo current rate limit info
+     * @param tokenBucket token rule info
+     */
+    private static void refill(XdsRateLimitInfo xdsRateLimitInfo, XdsTokenBucket tokenBucket) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: filltokenBucket, and remove it from consume method

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java (hunk 15 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: No judgment is required. If the XdsCoreService cannot be obtained, an exception is thrown.

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/XdsFlowControlConfig.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+     * xds flow control switch
+     */
+    private boolean enable;
+
+    /**
+     * whether to use secure protocol to invoke downstream service with xds route, example: http or https
+     */
+    @ConfigFieldKey("enabled-springcloud-xds-route-secure")
+    private boolean enabledSpringCloudXdsRouteSecure;
+
```
Comment (lilai23) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/XdsFlowControlConfig.java]: remove it

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/CommonConst.java [hunk 22]
```diff
@@ -269,6 +269,21 @@ public class CommonConst {
...
+
+    /**
+     * the connect for request address
+     */
+    public static final String CONNECT = ":";
+
+    /**
+     * Length of the cluster name after splitting
+     */
+    public static final int CLUSTER_NAME_SPLIT_LENGTH = 4;
```
Comment (lilai23) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/CommonConst.java]: move it into the class who use it

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+
+    /**
+     * route rule name
+     */
+    private String routeName;
+
+    /**
+     * request Address
+     */
+    private String address;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java:51]: request Address ip:port

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Circuit Breaker information
+ *
+ * @author zhp
+ * @since 2024-12-02
+ */
+public class XdsCircuitBreakerInfo {
```
Comment (lilai23) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java:29]: add comments for all fileds

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java (hunk 1 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java (hunk 2 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/core/match/XdsMatchManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java (hunk 3 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/handler/AbstractRequestHandler.java (hunk 4 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/handler/AbstractRequestHandler.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java (hunk 5 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 6 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java (hunk 7 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 8 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 9 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 10 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 11 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java (hunk 12 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java (hunk 13 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java (hunk 14 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/CircuitBreakerManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java (hunk 16 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java (hunk 17 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsFlowControlHandler.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java (hunk 18 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java (hunk 19 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java (hunk 20 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/ratelimit/RateLimitManager.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/XdsFlowControlConfig.java (hunk 21 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/XdsFlowControlConfig.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/CommonConst.java (hunk 22 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/config/CommonConst.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java (hunk 23 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/entity/FlowControlScenario.java:51]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 24 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java:29]: done

**Review by lilai23 — DISMISSED** [review]

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsHandler.java [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    /**
+     * get ServiceInstance of service name and cluster name
+     *
+     * @param serviceName service name
+     * @param clusterName cluster name
+     * @return Service Instance
+     */
+    public Set<ServiceInstance> getAllServerInstance(String serviceName, String clusterName) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsHandler.java]: getMatchedServiceInstance

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * Number of circuit breakers
+     */
+    private AtomicInteger circuitBreakerCount;
+
+    /**
+     * Constructor
+     */
+    public XdsCircuitBreakerInfo() {
+        this.localFailure = new ArrayDeque<>();
```
Comment (daizhenyu) [inline sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: use LinkedList

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsHandler.java (hunk 25 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/handler/XdsHandler.java]: done

File: sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java (hunk 26 above)
Comment (hanbingleixue) [reply sermant-plugins/sermant-flowcontrol/flowcontrol-common/src/main/java/io/sermant/flowcontrol/common/xds/circuit/XdsCircuitBreakerInfo.java]: done

**Review by daizhenyu — APPROVED** [review]

**Review by lilai23 — APPROVED** [review]

**Pull request merged**